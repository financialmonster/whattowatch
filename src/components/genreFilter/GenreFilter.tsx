import React, { FC } from 'react';
import { useSelector } from 'react-redux';
import cn from 'classnames';
import { Set } from 'immutable'

import * as filmsSelectors from 'domains/films/filmsSelectors';
import { isGenres } from 'utils';
import { useGenreFilter } from 'hooks/useGenreFilter';

export const GenreFilter: FC = () => {
    const genres = useSelector(filmsSelectors.selectGenres);
    const filmsError = useSelector(filmsSelectors.getFilmsError);
    const isFilmsFetching = useSelector(filmsSelectors.getIsFilmsFetching);

    const {filter, genreItemClickHander} = useGenreFilter();

    if (filmsError || isFilmsFetching) {
        return null;
    }

    let genresJSX: undefined | Set<JSX.Element>;
    
    if(isGenres(genres)) {
        genresJSX = genres.map((genre) => {
            const genreClass = cn(`catalog__genres-item`, {
                'catalog__genres-item--active': genre === filter
            });

            return (
                <li className={genreClass} key={ genre }>
                    <a className="catalog__genres-link" onClick={() => genreItemClickHander(genre)} title={ genre }>
                        { genre }
                    </a>
                </li>
            );
        });
    }

    const allGenresClass = cn(`catalog__genres-item`, {
        'catalog__genres-item--active': filter === `All genres`
    });

    return (
        <ul className="catalog__genres-list">
            <li className={allGenresClass}>
                <a className="catalog__genres-link" onClick={() => genreItemClickHander(`All genres`)}
                    title="All genres"
                >
                    All genres
                </a>
            </li>
            { genresJSX }
        </ul>
    );
};
