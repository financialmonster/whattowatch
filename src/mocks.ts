import { Map, fromJS, List } from 'immutable';

export const createStoreMock = (mode: string = `common`) => {
    return {
        promo: Map({
            promo: Map(),
            isPromoFetching: true,
            promoError: null
        }),
        films: Map({
            films: fromJS([{genre: `Thriller`, id: 1}, {genre: `Comedy`, id: 2}]),
            isFilmsFetching: mode === `detailed`,
            filmsError: null,
            filter: `All genres`
        }),
        auth: Map({
            user: null,
            authError: null,
            isAuthFetching: false,
            authStatusError: null,
            isAuthStatusFetching: false
        }),
        reviews: Map({
            reviews: List(),
            isReviewsFetching: true,
            reviewsError: null
        })
    }
}
