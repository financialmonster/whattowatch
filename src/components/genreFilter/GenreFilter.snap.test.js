import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { GenreFilter } from './GenreFilter';
import { createStoreMock } from 'mocks';

it(`GenreFilter: correctly rendered`, () => {
    const mockStore = configureStore();
    const store = mockStore(createStoreMock());

    const tree = renderer.create(
        <Provider store={store}>
            <GenreFilter />
        </Provider>        
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
