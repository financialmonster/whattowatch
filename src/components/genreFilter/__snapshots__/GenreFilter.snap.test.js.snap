// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreFilter: correctly rendered 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item catalog__genres-item--active"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
      style={
        Object {
          "cursor": "default",
        }
      }
      title=""
    >
      All genres
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
      style={Object {}}
      title="Comedy"
    >
      Comedy
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
      style={Object {}}
      title="Thriller"
    >
      Thriller
    </a>
  </li>
</ul>
`;
